File: MIDTERM_LAB_3/programming-2/assignment-02/MP08.java
/**
 * =====================================================
 * Student Name    : UNDUG, MOHAMMAD FAISSAL A.
 * Course          : Programming 2
 * Assignment      : MP08 - Filter Records by Keyword
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 18, 2026
 * GitHub Repo     : https://github.com/[your-username]/uphsd-cs-undug-mohammadfaissal
 *
 * Description:
 *   This program reads a CSV dataset file whose path is entered by the user.
 *   It then asks the user to type a keyword, and searches every field of every
 *   record for a case-insensitive match. All matching rows are printed in a
 *   formatted table along with a count of how many records were found.
 * =====================================================
 */

import java.io.*;
import java.util.*;

public class MP08 {

    // -- SECTION 1: Entry Point --------------------------------------------------
    // Collects the file path and search keyword from the user, then calls
    // the filter method to find and display all matching records.
    public static void main(String[] args) throws IOException {

        Scanner scanner = new Scanner(System.in);

        // Ask for the CSV file location first
        System.out.print("Enter the path to the CSV file: ");
        String filePath = scanner.nextLine().trim();

        // Ask for the keyword the user wants to search for
        System.out.print("Enter keyword to filter by: ");
        String keyword = scanner.nextLine().trim();

        filterByKeyword(filePath, keyword);
        scanner.close();
    }

    // -- SECTION 2: Filter Logic -------------------------------------------------
    // Reads every data row from the CSV and checks whether any field in that
    // row contains the keyword (case-insensitive). Matching rows are stored
    // in a list and then printed as a formatted table.
    static void filterByKeyword(String filePath, String keyword) throws IOException {

        // List to collect only the rows that match the keyword
        List<String[]> matches = new ArrayList<>();

        BufferedReader reader = new BufferedReader(new FileReader(filePath));
        String line;
        int lineNumber = 0;

        // Convert keyword to lowercase once so every comparison is case-insensitive
        String keywordLower = keyword.toLowerCase();

        while ((line = reader.readLine()) != null) {
            lineNumber++;

            // Skip the first 7 lines (metadata + column headers)
            if (lineNumber <= 7) continue;

            // Skip empty lines
            if (line.trim().isEmpty()) continue;

            String[] fields = parseCSVLine(line);

            // Only process rows that have enough columns
            if (fields.length <= 7) continue;

            // Check every field in this row for the keyword
            boolean rowMatches = false;
            for (String field : fields) {
                if (field.toLowerCase().contains(keywordLower)) {
                    rowMatches = true;
                    break; // No need to check remaining fields once a match is found
                }
            }

            // If any field matched, add the whole row to the results list
            if (rowMatches) {
                matches.add(fields);
            }
        }

        reader.close();

        // -- SECTION 3: Output ---------------------------------------------------
        // Display the filtered results in a clearly formatted table
        System.out.println();
        System.out.println("=".repeat(75));
        System.out.printf("  MP08 - RECORDS MATCHING KEYWORD: \"%s\"%n", keyword);
        System.out.println("  Student: UNDUG, MOHAMMAD FAISSAL A.");
        System.out.println("=".repeat(75));

        // Show a message and exit early if no records matched
        if (matches.isEmpty()) {
            System.out.printf("  No records found matching \"%s\".%n", keyword);
            System.out.println("=".repeat(75));
            return;
        }

        // Print the table header row
        System.out.printf("  %-25s %-12s %-35s %-6s %-6s%n",
                          "Candidate", "Type", "Exam", "Score", "Result");
        System.out.println("  " + "-".repeat(73));

        // Print each matching record
        for (String[] r : matches) {
            String name   = r.length > 0 ? r[0].trim() : "";
            String type   = r.length > 1 ? r[1].trim() : "";
            String exam   = r.length > 3 ? r[3].trim() : "";
            String score  = r.length > 6 ? r[6].trim() : "";
            String result = r.length > 7 ? r[7].trim() : "";

            // Truncate long exam names to keep table aligned
            if (exam.length() > 33) exam = exam.substring(0, 30) + "...";

            System.out.printf("  %-25s %-12s %-35s %-6s %-6s%n",
                              name, type, exam, score, result);
        }

        System.out.println("=".repeat(75));
        System.out.printf("  Records found: %d%n", matches.size());
        System.out.println("=".repeat(75));
    }

    // -- SECTION 4: CSV Line Parser ----------------------------------------------
    // Custom parser that handles quoted fields containing commas.
    // Iterates character by character, toggling a flag when inside quotes,
    // and only splits on commas that appear outside of quoted sections.
    static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean insideQuotes = false;

        for (char c : line.toCharArray()) {
            if (c == '"') {
                insideQuotes = !insideQuotes;
            } else if (c == ',' && !insideQuotes) {
                fields.add(current.toString());
                current.setLength(0);
            } else {
                current.append(c);
            }
        }

        fields.add(current.toString());
        return fields.toArray(new String[0]);
    }
}

File: MIDTERM_LAB_3/programming-2/assignment-02/mp08.js
/**
 * =====================================================
 * Student Name    : UNDUG, MOHAMMAD FAISSAL A.
 * Course          : Programming 2
 * Assignment      : MP08 - Filter Records by Keyword
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 18, 2026
 * GitHub Repo     : https://github.com/[your-username]/uphsd-cs-undug-mohammadfaissal
 * Runtime         : Node.js (run with: node mp08.js)
 *
 * Description:
 *   This program reads a CSV dataset file whose path is entered by the user.
 *   It then asks the user to type a keyword, and searches every field of every
 *   record for a case-insensitive match. All matching rows are printed in a
 *   formatted table along with a count of how many records were found.
 * =====================================================
 */

"use strict";

const fs       = require("fs");
const readline = require("readline");

// -- SECTION 1: CSV Line Parser -----------------------------------------------
// Splits a single CSV line into fields while handling quoted values that
// contain commas. Uses an insideQuotes boolean flag to track whether the
// current character is inside a quoted section or not.
function parseCSVLine(line) {
    const fields = [];
    let current = "";
    let insideQuotes = false;

    for (const c of line) {
        if (c === '"') {
            insideQuotes = !insideQuotes;
        } else if (c === "," && !insideQuotes) {
            fields.push(current);
            current = "";
        } else {
            current += c;
        }
    }

    fields.push(current);
    return fields;
}

// -- SECTION 2: Keyword Filter ------------------------------------------------
// Reads every data row in the CSV and checks whether any of its fields
// contain the search keyword (case-insensitive comparison using toLowerCase).
// Rows where at least one field matches are added to the results array.
// At the end, matching rows are printed as a formatted table.
function filterByKeyword(filePath, keyword) {

    const rawText    = fs.readFileSync(filePath, "utf8");
    const lines      = rawText.split(/\r?\n/);
    const keyLower   = keyword.toLowerCase(); // Pre-lowercase for faster comparisons
    const matches    = []; // Stores only the rows that match the keyword

    // Start reading data rows from index 7 (skip metadata + header lines)
    for (let i = 7; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;

        const fields = parseCSVLine(line);
        if (fields.length <= 7) continue;

        // Check if the keyword appears in any field of this row
        const rowMatches = fields.some(
            (field) => field.toLowerCase().includes(keyLower)
        );

        if (rowMatches) {
            matches.push(fields);
        }
    }

    // -- SECTION 3: Output ------------------------------------------------------
    // Display all matching records in a readable table format
    const line75 = "=".repeat(75);
    console.log();
    console.log(line75);
    console.log(`  MP08 - RECORDS MATCHING KEYWORD: "${keyword}"`);
    console.log("  Student: UNDUG, MOHAMMAD FAISSAL A.");
    console.log(line75);

    // If nothing matched, tell the user and stop
    if (matches.length === 0) {
        console.log(`  No records found matching "${keyword}".`);
        console.log(line75);
        return;
    }

    // Print column headers
    console.log(
        "  " +
        "Candidate".padEnd(25) +
        "Type".padEnd(12) +
        "Exam".padEnd(35) +
        "Score".padEnd(6) +
        "Result"
    );
    console.log("  " + "-".repeat(73));

    // Print each matching row
    matches.forEach((r) => {
        let name   = (r[0] || "").trim();
        let type   = (r[1] || "").trim();
        let exam   = (r[3] || "").trim();
        let score  = (r[6] || "").trim();
        let result = (r[7] || "").trim();

        // Truncate exam names that are too long to keep columns tidy
        if (exam.length > 33) exam = exam.substring(0, 30) + "...";

        console.log(
            "  " +
            name.padEnd(25) +
            type.padEnd(12) +
            exam.padEnd(35) +
            score.padEnd(6) +
            result
        );
    });

    console.log(line75);
    console.log(`  Records found: ${matches.length}`);
    console.log(line75);
}

// -- SECTION 4: Program Entry Point -------------------------------------------
// Uses readline to ask for the file path first, then the search keyword.
// Both inputs are collected before the filter function is called.
const rl = readline.createInterface({
    input:  process.stdin,
    output: process.stdout
});

// Ask for file path first, then ask for keyword inside the callback
rl.question("Enter the path to the CSV file: ", (filePath) => {
    rl.question("Enter keyword to filter by: ", (keyword) => {
        rl.close();
        filterByKeyword(filePath.trim(), keyword.trim());
    });
});
